### DISCUSSION TITLE
Middleware using shutdown event, can it be ported to use lifespan?

### DISCUSSION THREAD

**Discussion by voltagex** [fastapi/fastapi#11872] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
N/A
```

### Description

https://github.com/Jijun/fastapi-cprofile uses shutdown as a hook to dump performance statistics to a file.

I don't see any examples of "attaching" to lifespan from within middleware - so I'm interested in working out if it's possible. I think this old middleware is still useful.

### Operating System

Linux

### FastAPI Version

-

### Pydantic Version

-

### Python Version

-

**Comment by YuriiMotov** [comment, marked as answer]
Not sure it's a good idea, but you can call shutdown handlers manually in the lifespan function:
```
from contextlib import asynccontextmanager
from fastapi import FastAPI

@asynccontextmanager
async def lifespan(app: FastAPI):
    yield
    for on_shutdown in  app.router.on_shutdown:
        await on_shutdown()

app = FastAPI(lifespan=lifespan)

async def get_profiler_result():
    print("On shutdown has been called")

app.add_event_handler("shutdown", get_profiler_result)


@app.get("/")
def index():
    pass

```

**Comment by hitokiri** [comment]
@YuriiMotov 
I would advise against this approach, as it does not account for all the scenarios handled by the previous one. For instance, it lacks support for combinations with the decorator:
@repeat_every(seconds=INTERVAL, wait_first=True), just to give one example.

**Comment by SMAntony** [comment]
Any alternative for this?

**Reply by YuriiMotov** [reply]
I think alternative is to ask maintainers of those libraries to implement or document the way of using them with `lifespan`
